### DISCUSSION TITLE
Deploying Index changes without downtime

### DISCUSSION THREAD

**Discussion by venick** [ravendb/ravendb#12898] [Q&A] [answered]
Hi,

We have an ASP.NET application running on RavenDB Cloud that needs to be be deployed with zero downtime. These deployments sometimes involve changes to the indexes.

### Example Scenario
An exiting index needs the be modified.

**Old version**
```
docs.Employees.Select(entity => new {
    Name = entity.Name,
    ShortName = entity.Name.Split(' ')[0]
})
```
**New version**
```
docs.Employees.Select(entity => new {
    FullName = entity.Name,
    ShortName = entity.Name.Split(' ')[1],
    StartDate = entity.StartDate
})
```
In this example we are adding a new field to be indexed, changing the `ShortName` to use the surname instead of first name and changed the `Name` field to `FullName`.

There are 1 million records, and re-indexing takes 5 mins.

### Problem
The old build is expecting the index to have a `Name` field so if the new index is swapped in too early and queries that filter or sort by `Name` will fail.
The new build is expecting a `StartDate` so if the old index is still active then sorting by `StartDate` will fail.

Since the index takes 5 mins to rebuild, if we apply to index changes _after_ the new build is deployed, then there will be a 5 min window where attempts to filter/sort by `StartDate` will throw an exception. However if we wait for the index to rebuild before deploying the new build, then filtering/sorting on `Name` will throw an exception.

### Question
How can we coordinate so that the new index is swapped in at the same moment that the new build goes live?
Are there any versioning features on indexes we can use, where two versions of the same index can exist in parallel during deployments?
Can auto-swap be disabled on indexes that are re-indexing in parallel until we are ready?

**Comment by mcrio** [comment, marked as answer]
Hi, just a side question, the folks from RavenDB will be able to answer precisely about index swapping: Can you instead of updating the old index, create a new one which the new builds will use, and once the whole system is updated you go ahead and remove the old index as part of the cleanup?

**Reply by ayende** [reply]
That would be my recommendation as well. The complexity of coordinating zero downtime index replacement in a distributed system is *hard*.
Having `Index/V1` and `Index/V2` instead, where you can deploy in your own time, verify that it is done, and then deploy your code is much nicer.

There is also the advantage that you can run both versions of your application at the same time.

**Reply by venick** [reply]
Thanks guys that's great help.

I implemented a solution which keeps both indexes in parallel in RavenDB (until the the old no longed needed) but still only ever refers to the single index class in code. I did this by overriding the `IndexName` property to add a version number to the index name.

In case anyone else is interested my solution looks like this:
```
public abstract class VersionedAbstractIndexCreationTask<TDocument, TReduceResult> : AbstractIndexCreationTask<TDocument, TReduceResult>
{
	public abstract int Version { get; }

	public override string IndexName
	{
		get
		{
			return Version > 0
				? $"{GetType().Name.Replace('_', '/')}/{Version}"
				: GetType().Name.Replace('_', '/');
		}
	}
}
```
I now implement this abstract class instead of `AbstractIndexCreationTask` when I want to add versioning to the index and increment the `Version` property whenever I make changes.
